File: KMACTF 2024/ZKP for QNR/python/verifier.py
import random
from typing import List, Tuple, Union
from dataclasses import dataclass, field

@dataclass
class Verifier:
    x: int
    y: int
    rs: List[List[Tuple[int, int]]] = field(default_factory=list)
    pairs: List[List[Tuple[int, int]]] = field(default_factory=list)
    num_rounds: int = 0

    def gen_w(self, bit: int, r: int) -> int:
        if bit == 0:
            return pow(r, 2, self.x)
        else:
            return (pow(r, 2, self.x) * self.y) % self.x

    def gen_pairs(self) -> List[Tuple[int, int]]:
        rr = []
        pair = []
        for _ in range(self.num_rounds):
            r1 = self.rand_below(self.x)
            r2 = self.rand_below(self.x)
            a = pow(r1, 2, self.x)
            b = (pow(r2, 2, self.x) * self.y) % self.x

            pair.append((a, b))
            rr.append((r1, r2))
        self.rs.append(rr)
        self.pairs.append(pair)
        return pair

    def respond(self, bit: int, list_i: List[int], r: int) -> List[Union['RpTuple', 'RpNumber']]:
        v = []
        for i, a in enumerate(list_i):
            if a == 0:
                vj = self.rs[-1][i]
                v.append(RpTuple(vj[0], vj[1]))
            elif a == 1:
                if bit == 0:
                    vj = (r * self.rs[-1][i][0]) % self.x
                elif bit == 1:
                    vj = (r * self.rs[-1][i][1] * self.y) % self.x
                v.append(RpNumber(vj))
            else:
                raise ValueError("Invalid input")
        return v

    def get_r(self) -> int:
        return self.rand_below(self.x)

    @staticmethod
    def rand_below(n: int) -> int:
        """Generate a random integer less than n."""
        bit_length = n.bit_length()
        while True:
            y = random.getrandbits(bit_length)
            if y < n:
                return y

@dataclass
class RpTuple:
    a: int
    b: int

@dataclass
class RpNumber:
    n: int

File: KMACTF 2024/ZKP for QNR/src/verifier.rs
use crate::utils::*;
use num_bigint::BigUint;
use serde_derive::Serialize;
pub struct Verifier {
    pub x: BigUint,
    pub y: BigUint,
    rs: Vec<Vec<(BigUint, BigUint)>>,
    pub pairs: Vec<Vec<(BigUint, BigUint)>>,
    pub num_rouns: usize,
}

#[derive(Serialize, Debug)]
pub enum Rp {
    Tuple(BigUint, BigUint),
    Number(BigUint),
}

impl Verifier {
    pub fn new(x: BigUint, y: BigUint, n: usize) -> Self {
        Verifier {
            x,
            y,
            rs: Vec::new(),
            pairs: Vec::new(),
            num_rouns: n,
        }
    }

    pub fn gen_w(&self, bit: i32, r: &BigUint) -> BigUint {
        if bit == 0 {
            r.modpow(&BigUint::from(2_u32), &self.x)
        } else {
            r.modpow(&BigUint::from(2_u32), &self.x) * &self.y % &self.x
        }
    }

    pub fn gen_pairs(&mut self) -> Vec<(BigUint, BigUint)> {
        let mut rr: Vec<(BigUint, BigUint)> = Vec::with_capacity(self.num_rouns);
        let mut pair: Vec<(BigUint, BigUint)> = Vec::with_capacity(self.num_rouns);
        for _ in 0..self.num_rouns {
            let r1 = rand_below(&self.x);
            let r2 = rand_below(&self.x);
            let a = r1.modpow(&BigUint::from(2_u32), &self.x);
            let b = r2.modpow(&BigUint::from(2_u32), &self.x) * &self.y % &self.x;

            pair.push((a, b));
            rr.push((r1, r2));
        }
        self.rs.push(rr.clone());
        self.pairs.push(pair.clone());
        pair
    }

    pub fn respond(&self, bit: &u32, list_i: Vec<u32>, r: &BigUint) -> Vec<Rp> {
        let mut v: Vec<Rp> = Vec::with_capacity(self.num_rouns);
        for (i, a) in list_i.iter().enumerate() {
            match (a, bit) {
                (&0, _) => {
                    let vj = &self.rs[self.rs.len() - 1][i];
                    v.push(Rp::Tuple(vj.0.clone(), vj.1.clone()));
                }
                (1, &0_u32) => {
                    let vj = r * &self.rs[self.rs.len() - 1][i].0 % &self.x;
                    v.push(Rp::Number(vj));
                }
                (1, &1_u32) => {
                    let vj = r * &self.rs[self.rs.len() - 1][i].1 * &self.y % &self.x;
                    v.push(Rp::Number(vj));
                }
                _ => panic!("Invalid input"),
            }
        }
        v
    }

    pub fn get_r(&self) -> BigUint {
        rand_below(&self.x)
    }
}
